android/src/main/java/com/swmansion/gesturehandler/core/FlingGestureHandler.kt

```diff
@@ -35,26 +34,74 @@ class FlingGestureHandler : GestureHandler<FlingGestureHandler>() {
     handler!!.postDelayed(failDelayed, maxDurationMs)
   }
 
-  private fun tryEndFling(event: MotionEvent) = if (
-    maxNumberOfPointersSimultaneously == numberOfPointersRequired &&
-    (
-      direction and DIRECTION_RIGHT != 0 &&
-        event.rawX - startX > minAcceptableDelta ||
-        direction and DIRECTION_LEFT != 0 &&
-        startX - event.rawX > minAcceptableDelta ||
-        direction and DIRECTION_UP != 0 &&
-        startY - event.rawY > minAcceptableDelta ||
-        direction and DIRECTION_DOWN != 0 &&
-        event.rawY - startY > minAcceptableDelta
+  private fun tryEndFling(event: MotionEvent): Boolean {
+    data class SimpleVector(val x: Double, val y: Double)
+
+    fun toSafeNumber(unsafe: Double): Double = if (unsafe.isFinite()) unsafe else 0.0
+
+    fun toUnitVector(vec: SimpleVector): SimpleVector {
+      val magnitude = hypot(vec.x, vec.y)
+      // toSafeNumber protects against division by zero
+      return SimpleVector(
+        toSafeNumber(vec.x / magnitude),
+        toSafeNumber(vec.y / magnitude)
       )
-  ) {
-    handler!!.removeCallbacksAndMessages(null)
-    activate()
-    true
-  } else {
-    false
-  }
+    }
+
+    fun compareSimilarity(
+      vecA: SimpleVector,
+      vecB: SimpleVector
+    ): Double {
+      // inputs have to be unit vectors
+      // returns scalar on range from -1.0 to 1.0
+      return vecA.x * vecB.x + vecA.y * vecB.y
+    }
+
+    fun compareAlignment(
+      vec: SimpleVector,
+      directionVec: SimpleVector,
+      direction: Int
+    ): Boolean =
+      compareSimilarity(vec, directionVec) > minDirectionAlignment &&
+        (this.direction and direction != 0)
+
+    val velocityTracker = VelocityTracker.obtain()
+    addVelocityMovement(velocityTracker, event)
+    velocityTracker!!.computeCurrentVelocity(1000)
+
+    val velocityVector = SimpleVector(
+      velocityTracker.xVelocity.toDouble(),
+      velocityTracker.yVelocity.toDouble()
+    )
+
+    val velocityUnitVector = toUnitVector(velocityVector)
 
+    val alignmentList = arrayOf(
+      compareAlignment(velocityUnitVector, SimpleVector(-1.0, 0.0), DIRECTION_LEFT),
+      compareAlignment(velocityUnitVector, SimpleVector(1.0, 0.0), DIRECTION_RIGHT),
+      compareAlignment(velocityUnitVector, SimpleVector(0.0, -1.0), DIRECTION_UP),
+      compareAlignment(velocityUnitVector, SimpleVector(0.0, 1.0), DIRECTION_DOWN)
+    )
+
+    velocityTracker.recycle()
+
+    val totalVelocity = hypot(velocityVector.x, velocityVector.y)
+
+    val isAligned = alignmentList.any { it }
```

Comment:
I feel like this should be its own class (minus the velocity stuff at the end). And then `Simple` can be dropped from the name, since it will no longer be that simple 😄.

Instead of creating directional vectors for every check, they can be static fields of the class (which are achieved using [`companion object` in Kotlin](https://kotlinlang.org/docs/object-declarations.html#companion-objects)).

And I'd move it outside the `tryEndFling` so it's inner class of Fling or move it to a separate file altogether.

Reply:
I've separated most of the logic into a separate Vector class, please let me know if this is what You had in mind, and if i should place it in a separate file, or is it fine to keep it where it is :)